sqlalchemy-stubs/engine/interfaces.pyi

```diff
@@ -1,214 +1,315 @@
 from typing import Any
+from typing import Dict
+from typing import List
+from typing import Mapping
 from typing import Optional
+from typing import Sequence
+from typing import Tuple
+from typing import Type
+from typing import Union
 
-from .url import URL as URL
-from .. import util as util
+from typing_extensions import Protocol
+
+from .base import Connection
+from .base import Engine
+from .cursor import CursorFetchStrategy
+from .cursor import CursorResult
+from .url import URL
+from ..exc import StatementError
 from ..sql.compiler import Compiled as Compiled
+from ..sql.compiler import IdentifierPreparer
 from ..sql.compiler import TypeCompiler as TypeCompiler
-from ..util import compat as compat
+from ..sql.type_api import TypeEngine
+
+class _OnConnect(Protocol):
+    def __call__(self, connection: Connection) -> None: ...
+
+class _DBAPIConnection(Protocol):
+    def close(self) -> None: ...
+    def commit(self) -> None: ...
+    def cursor(self) -> _DBAPICursor: ...
```

Comment:
`rollback()` is missing. https://www.python.org/dev/peps/pep-0249/#rollback

mmm, reading at the definition it says:
>This method is optional since not all databases provide transaction support

is it possible to define that a method is optional?

Reply:
Not exactly, so I'm gonna define a second protocol that inherits from this and has `rollback()`. I'll do the same with the optional methods in `_DBAPICursor` below.

Reply:
personally I would be pragmatic on this one and just define a rollback method in `_DBAPIConnection`.
I think that the number of dbapi that do not support it is very small compared to the total